### PR TITLE
[WIP] JUnit 5

### PR DESCRIPTION
**Pull request by sormuras**
_in progress..._

- [x] Use JUnit Platform at test runtime
- [x] Migrate single test to use JUnit Jupiter API
- [ ] Don't over-use the Jupiter API, tune it down a bit

Related to #603 and https://github.com/junit-team/junit5/issues/1582

### CODE REVIEW

**Comment by JakeWharton** [comment]
JavaPoet is imported as source into Google's monorepo so I'll have to check as to whether JUnit 5 is supported or not. Or maybe @ronshapiro knows immediately off-hand (I'm inclined to think no).

My favorite quote from the linked issue is:

> they are two separate frameworks

JUnit 5 was named incorrectly. This is an entirely different framework that should be vetted on its own merits compared to JUnit 4. It is decidedly **not** an upgrade.

So we should evaluate it against other new frameworks and also against doing nothing.

**Comment by ronshapiro** [comment]
It's not, but we also don't run JavaPoet's tests (for a silly reason, I think at the time none of the tests had `@RunWith` and our runner didn't support detecting them (over JUnit3 tests)).

But please don't make a decision based on Google's eng structure (especially for tests!)

**Comment by swankjesse** [comment]
Yep. Because JUnit 4 and JUnit 5 don’t interoperate nicely I think we’re better off waiting until all of our dependent projects have upgraded. In particular, we should wait for compile-testing and Mockito to support JUnit 5.

File: src/test/java/com/squareup/javapoet/UtilTest.java [hunk 1]
```diff
@@ -15,61 +15,87 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.DynamicContainer.dynamicContainer;
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
 
-import org.junit.Test;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.DynamicNode;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestFactory;
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/UtilTest.java]: Going from 2 imports to 9 makes me want to back away from Jupiter forever.

File: pom.xml [hunk 2]
```diff
@@ -69,6 +73,18 @@
...
     </dependency>
+    <dependency>
+      <groupId>org.junit.vintage</groupId>
+      <artifactId>junit-vintage-engine</artifactId>
+      <version>${junit.vintage.version}</version>
+      <scope>runtime</scope> <!-- test-runtime -->
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId> <!-- junit-jupiter-api, Surefire is not smart enough, yet. -->
```
Comment (swankjesse) [inline pom.xml]: Blech. They say one of Jupiter’s benefits is better integration with test runners. But I see that it isn’t integrated with Maven yet. That doesn’t inspire confidence.

**Comment by sormuras** [comment]
Mockito 2.22.0 already provides a `MockitoExtension`: https://static.javadoc.io/org.mockito/mockito-junit-jupiter/2.22.0/org/mockito/junit/jupiter/MockitoExtension.html

File: src/test/java/com/squareup/javapoet/UtilTest.java (hunk 1 above)
Comment (sormuras) [reply src/test/java/com/squareup/javapoet/UtilTest.java]: +7 more imports to go from

image

to

image

Better reporting needs some imports.

But as I stated in the original description text, I'll rework this commit to only show minimal changes.

File: pom.xml (hunk 2 above)
Comment (sormuras) [reply pom.xml]: > Blech.

Hey, I understand that. (-:

Come over and help improving Surefire!

Or try https://github.com/sormuras/junit-platform-maven-plugin which already "is smart enough" to add required runtime dependencies.

Comment (sormuras) [reply pom.xml]: Surefire 2.22.2 or the upcoming 3.0.0-M4 do help in this configuration ... "Blech". I'll upgrade the PR ... soon.
